examples/applications/plot_cyclical_feature_engineering.py

```diff
@@ -0,0 +1,534 @@
+# to train machine learning models. Instead it we simplify the representation
+# by collapsing those into the `"rain"` category.
+#
+X["weather"][X["weather"] == "heavy_rain"] = "rain"
+# %%
+X["weather"].value_counts()
+
+# %%
+# As expected season variable is well balanced:
+#
+X["season"].value_counts()
+
+# %%
+# Time-based Cross-Validation Splits
+# ----------------------------------
+#
+# Since the dataset is a time-ordered event log (hourly demand), we will use a
+# time-sensitive cross-validation splitter to evaluate our demand forecasting
+# model as realistically as possible. We use a gap of 2 days between the train
+# and test side of the splits. We also limit the training set size to make the
+# performance of the CV folds more stable.
+#
+# 1000 test datapoints should be enough to quantify the performance of the
+# model. This reprent a bit less than a month and a half of contiguous test
+# data:
+
+from sklearn.model_selection import TimeSeriesSplit
+
+ts_cv = TimeSeriesSplit(
+    n_splits=5, gap=48, max_train_size=10000, test_size=1000,
+)
+
+# %%
+# Let us manually inspect the various splits to check that the
+# `TimeSeriesSplit` works as we expect, starting with the first split:
+all_splits = list(ts_cv.split(X, y))
+train_0, test_0 = all_splits[0]
+
+# %%
+X.iloc[test_0]
+
+# %%
+X.iloc[train_0]
+
+# %%
+# We now inspect the last split:
+train_4, test_4 = all_splits[4]
+
+# %%
+X.iloc[test_4]
+
+# %%
+X.iloc[train_4]
+
+# %%
+# All is well. We are now ready to do some predictive modeling!
+#
+# Gradient Boosting
+# -----------------
+#
+# Gradient Boosting Regression with decision trees is often flexible enough
+# to efficiently handle heteorogenous tabular data with a mix of categorical
+# and numerical features as long as the number of samples is large enough.
+#
+# Here we do minimal ordinal encoding for the categorical variables and then
+# let the model know that it should treat those a categorical variables using
+# a dedicated split rule.
+#
+# The numerical variable need no preprocessing and for the sake of simplicity
+# we only try the default hyper-parameters for this model:
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import OrdinalEncoder
+from sklearn.compose import ColumnTransformer
+from sklearn.ensemble import HistGradientBoostingRegressor
+from sklearn.model_selection import cross_validate
+
+
+categorical_columns = [
+    "weather",
+    "season",
+    "holiday",
+    "workingday",
+]
+categories = [
+    ["clear", "misty", "rain", "heavy_rain"],
+    ["spring", "summer", "fall", "winter"],
+    ["False", "True"],
+    ["False", "True"],
+]
+ordinal_encoder = OrdinalEncoder(categories=categories)
+
+
+gbrt_pipeline = make_pipeline(
+    ColumnTransformer(
+        transformers=[
+            ("categorical", ordinal_encoder, categorical_columns),
+        ],
+        remainder="passthrough",
+    ),
+    HistGradientBoostingRegressor(
+        categorical_features=range(4),
```

Comment:
Maybe fix the random_state for reproducibility?

Reply:
In my experience this estimator is deterministic in this case. Only feature binning is non-deterministic when n_samples is too large which is not the case in this example: n_samples ~= 17k while the subsampling threshold is 20k.